platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/heartbeats/HeartbeatProtocol.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.heartbeats;
+
+import com.swirlds.base.ArgumentUtils;
+import com.swirlds.common.system.NodeId;
+import com.swirlds.platform.Connection;
+import com.swirlds.platform.network.ByteConstants;
+import com.swirlds.platform.network.NetworkMetrics;
+import com.swirlds.platform.network.NetworkProtocolException;
+import com.swirlds.platform.network.protocol.Protocol;
+import com.swirlds.platform.network.unidirectional.UnidirectionalProtocols;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+
+/**
+ * Sends a heartbeat to the other node and measures the time it takes to receive a response.
+ */
+public class HeartbeatProtocol implements Protocol {
+    /**
+     * ID of the peer
+     */
+    private final NodeId otherId;
+
+    /**
+     * The last time a heartbeat protocol was executed
+     */
+    private Instant lastHeartbeatTime = Instant.MIN;
+
+    /**
+     * The period at which the heartbeat protocol should be executed
+     */
+    private final Duration heartbeatPeriod;
+
+    /**
+     * Network metrics, for recording roundtrip heartbeat time
+     */
+    private final NetworkMetrics networkMetrics;
+
+    /**
+     * Constructor
+     *
+     * @param otherId         ID of the peer
+     * @param heartbeatPeriod The period (milliseconds) at which this protocol should execute
+     * @param networkMetrics  Network metrics, for recording roundtrip heartbeat time
+     */
+    public HeartbeatProtocol(
+            final @NonNull NodeId otherId,
+            final @NonNull Duration heartbeatPeriod,
+            final @NonNull NetworkMetrics networkMetrics) {
+
+        this.otherId = ArgumentUtils.throwArgNull(otherId, "otherId");
+        this.heartbeatPeriod = ArgumentUtils.throwArgNull(heartbeatPeriod, "heartbeatPeriod");
+        this.networkMetrics = ArgumentUtils.throwArgNull(networkMetrics, "networkMetrics");
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Returns true if the last heartbeat protocol was started more than {@link #heartbeatPeriod} ago
+     */
+    @Override
+    public boolean shouldInitiate() {
+        return Duration.between(lastHeartbeatTime, Instant.now()).compareTo(heartbeatPeriod) >= 0;
```

Comment:
Instead of calling `Instant.now()`, pass in an instance of `Time`. It makes this class unit testable

Reply:
done